core/src/common.cpp

```diff
@@ -362,6 +364,448 @@ namespace irods::http
+		// See RFC 7515 Section 4.1.4
+		if (_jwt.has_key_id()) {
+			auto key_id{_jwt.get_key_id()};
+			if (_jwks.has_jwk(key_id)) {
+				auto jwk{_jwks.get_jwk(key_id)};
+				add_asymmetric_alg_from_jwk(_verifier, jwk, alg);
+
+				return _verifier;
+			}
+			logging::warn("{}: Could not find the desired [kid] in the JWKs list.", __func__);
+		}
+		// We cannot pick out the specific key used, go through entire list of JWKs
+
+		// The first two characters of 'alg' should give us
+		// enough information to get the algorithm 'family'
+		const auto algorithm_family{alg.substr(0, 2)};
+
+		// 'kty' string to search for in JWK
+		// The only valid values are 'EC', 'RSA', and 'oct'
+		// See (JWA Section 6.1) for the table of valid values
+		std::string search_string;
+
+		// RSA family (JWA Section 3.1)
+		if (algorithm_family == "RS" || algorithm_family == "PS") {
+			// 'kty' string for RSA (JWA Section 6.1)
+			search_string = "RSA";
+		}
+		// EC family (JWA Section 3.1)
+		else if (algorithm_family == "ES") {
+			// 'kty' string for Elliptic Curve (JWA Section 6.1)
+			search_string = "EC";
+		}
+		// Symmetric algo (JWA Section 3.1)
+		else if (algorithm_family == "HS") {
+			add_symmetric_alg(_verifier, alg);
+			return _verifier;
+		}
+		// Not a valid or supported algorithm
+		else {
+			logging::error("{}: 'alg' of [{}] is unsupported.", __func__, alg);
+			return _verifier;
+		}
+
+		// Go through entire key set
+		std::for_each(
+			std::cbegin(_jwks),
+			std::cend(_jwks),
+			[fn = __func__, &_verifier, &alg, &search_string](const auto& _jwk) -> void {
+				// Check the optional claims first
+			    // Skip JWK if 'use' is not for signing 'sig'
+			    // See JWK Section 4.2
+				if (_jwk.has_use() && _jwk.get_use() != "sig") {
+					logging::trace("{}: JWK not a signing key, ignoring.", fn);
+					return;
+				}
+				// JWK might have 'key_ops', try to select keys based off of this
+			    // See JWK Section 4.3
+				if (_jwk.has_key_operations() && !_jwk.get_key_operations().contains("verify")) {
+					logging::trace("{}: JWK not a key used for verification, ignoring.", fn);
+					return;
+				}
+
+				if (_jwk.has_algorithm()) {
+					// Add the algorithm if 'alg' matches desired.
+					if (_jwk.get_algorithm() == alg) {
+						add_asymmetric_alg_from_jwk(_verifier, _jwk, alg);
+						return;
+					}
+
+					logging::trace("{}: JWK [alg] does not match JWT [alg], ignoring.", fn);
+					return;
+				}
+
+				// Fallback to required claim
+				if (_jwk.has_key_type()) {
+					// Extract the 'kty' of the JWK, compare to desired 'kty'
+					if (_jwk.get_key_type() == search_string) {
+						add_asymmetric_alg_from_jwk(_verifier, _jwk, alg);
+						return;
+					}
+
+					logging::trace("{}: JWK [kty] does not match JWT desired [kty], ignoring.", fn);
+					return;
+				}
+
+				logging::error("{}: Invalid JWK, missing [kty] claim. Ignoring.", fn);
+				return;
+			});
+
+		// Allow for chaining
+		return _verifier;
+	}
```

Comment:
`} // add_algorithms_to_verifier`

Please do the same for the other functions.

Reply:
Done.